Deploy to

kalbir opened the pull request:
A joint effort from @tatyree and I -- writing a sinatra app that labels issues with the label `support team` after they are created.

There is quite a lot here. I suggest that the best review is:

- [x] to create a new namespace on Kubernetes for the `production` version of the app and deploying it there. If you can do that then the `README` is in decent shape 
- [x] look over `webhook.rb` to understand how the app handles the incoming payload from github and processes it
- [x] look over the `Dockerfile` (which is really just an expression of me cursing that there are not enough boilerplate dockerfiles or best practices ) and see if that makes sense/there are any changes 

I appreciate there aren't many Rubyists on our team so if you want to do a walkthrough with me, I'm happy to.

The app is currently running in live, labelling tickets on the `cloud-platform` repo, but it is running from the `support-labelling-webhook-dev` namespace.

tatyree committed:
WIP: Add most basic sinatra app for testing

tatyree committed:
Add puma as webserver

Mainly for efficiency and consistency-this is generally how we run all
our rack-based services on heroku.

tatyree committed:
Update ruby version to 2.5.1

Heroku default was `2.4.4`.  My rbenv was out-of-date so it seemed
sensible to bring the ruby version up the Heroku's most current version
while updating by local environment.

tatyree committed:
Add puma as webserver

tatyree committed:
WIP: Add most basic sinatra app for testing

Creates and extends a sinatra app that will be a webhook endpoints for a
GitHub webhook. The idea is to define an endpoint `/webhook` that takes
in issue data from Github when a new issue is created and adds a label
to issues that match a particular regex.

To make this work we need to define a regex, setup access to the github
API, parse the incoming json payload and add the label. This code does
that.

kalbir committed:
Make Puma run by default

We don't want to run sinatra using the default rack webserver but
instead want to use Puma. Puma is installed (check out the gemfile) but
we can use this little bit of config in the main sinatra file
`webhook.rb` to make sure Puma is used everytime you start the
application.

I've also added a little bit to the readme that reflects this way of
doing things.

kalbir committed:
Add a Dockerfile for the app

We had been deploying onto Heroku and so just needed a buildpack (ah,
the good old days!). Now we need a dockerfile so that we can build and
deploy a container to Kubernetes. This is my first go, based on some I
found on the internet. There are some tidy ups we can do, for instance:

* use multistage builds
* move some of the commands into a startup script
* use a non root user

but this is a reasonable start I think.

kalbir committed:
Add Kubernetes deployment files

These are based on the cloud platform user guide manual deployment
method and adding a secret docs. Those two docs don't quite cohere so I
had to smooth them together a bit in adding the env vars to the
deployment file.

kalbir committed:
Make the app work on Kubernetes

There is a key change that needs to be made to make sure the app can run
on kubernetes which is in `webhook.rb` here. We need to make sure the
app binds to `0.0.0.0` not just to `localhost` which is [the
default](http://sinatrarb.com/configuration.html) hostname (see the
`:bind` section of that page).

I also added the docker command `EXPOSE` as that looks like best
practice.

kalbir committed:
Tidy up deployment files

A few changes to make the deployment files use better names and the
latest version of the docker image. Main things are:

* the app is now called "support-ticket-labelling"
* the ingress is on port 80 and the app is on 4567, the service
translates between the two
* I made it clearer that the secret defined here is not a "real* secret
- I'm going to put in a proper secret via a different method, not adding
it into the repo

kalbir committed:
Update the README

This is the outline of how the app works and gets deployed, bearing in
mind that I haven't finished sorting out the secret bit so I haven't
tested it working on Kubernetes end to end. But gives a reasonable idea
of how to get started.

kalbir committed:
Used a later version of bundler

kalbir committed:
Update to v0.6

This is the version that is live. It labels with the 'support team'
label, 0.5 got everything working apart from that.

kalbir committed:
Add full deployment instructions

I've updated the README with the full instructions to get the app
deployed in its current state. This involved making a couple of small
tidy up changes (delete `secrets.yaml` as we don't need it and get rid
of some comments in `webhook.rb`)

The things that I've left as is here but that might merit a bit of
change are:

* the "live" app deploys to the `-dev` namespace. The reason I have left
this as is is because I haven't had time to move the ECR repo. I'll
probably do that later but it is working now from `dev`.
* we are deploying the app manually, we could quite easily have a circle
CI job to do this, but is it worth it?

kalbir committed:
Do not commit tool versions file

[asdf](https://github.com/asdf-vm/asdf) the version manager I use puts a
`.tool-versions` file into the repo. I don't want to commit this as we
have the info in the Gemfile and I don't want to have to keep the two in
sync -- `.tool-versions` is for local sync.

jasonBirchall commented on README.md:
```diff
@@ -1,2 +1,98 @@
 A webhook for github to ensure that new support issues are labelled correctly when added by non-WebOps team members.
+
+A webhook for github to ensure that new support issues are labelled
+correctly when added by non-Webops team members.
```
Duplicate block of text.

jasonBirchall reviewed (changes requested):
Just a small request for change. App deployed following readme and DockerFile looks good.

kalbir committed:
Delete duplicate paragraph

Also change "WebOps" to "Cloud Platform", as that is what we are actually called.

kalbir commented:
@jasonBirchall Made that change you requested 👍
